### PR TITLE
Update `_mut_iter_equal_skeleton` method

### PR DESCRIPTION
**Pull request by TwoPair**
This code is tested based on `list_tests.py::CommonTest::test_remove`,
except for lines `352~358` that seem related to `str` test.

### CODE REVIEW

**Comment by youknowone** [comment]
line 352 is not related to string. try to print its type like `print(type(self.type2test))` that's list type. e.g. `list`.
`list('abcdefghcij')` is `['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'c', 'i', 'j']`

**Comment by youknowone** [comment]
By watching CI, it is causing dead-lock. it didn't finish during 4 hours.

**Comment by TwoPair** [comment]
Now, two list method `index`, `remove` works normally by fixing `_mut_iter_equal_skeleton` method.

File: vm/src/sequence.rs [hunk 1]
```diff
@@ -81,91 +72,10 @@ pub trait MutObjectSequenceOp<'a> {
...
-                            }
-                        }
-                    }
+                let eq = {
+                    let elem = elem.clone();
+                    drop(guard);
+                    elem.rich_compare_bool(needle, PyComparisonOp::Eq, vm)?
                 };
 
                 if eq {
```
Comment (Snowapril) [inline vm/src/sequence.rs:81]:
```suggestion
                let elem = elem.clone();
                drop(guard);
                
                if elem.rich_compare_bool(needle, PyComparisonOp::Eq, vm)? {
```

**Comment by Snowapril** [comment]
Nice catch! The changed codes seem totally fine to me. Please check unexpected success in `test_list`, `test_deque` tests.

File: Lib/test/test_list.py [hunk 2]
```diff
@@ -231,13 +231,13 @@ def __eq__(self, other):
         lst = [X(), X()]
         X() in lst
 
-    # TODO: RUSTPYTHON
-    @unittest.expectedFailure
+    # # TODO: RUSTPYTHON
+    # @unittest.expectedFailure
     def test_count(self):  # XXX: RUSTPYTHON; the method also need to be removed when done
         super().test_count()
```
Comment (fanninpm) [inline Lib/test/test_list.py]:
```suggestion
```

File: Lib/test/test_list.py [hunk 3]
```diff
@@ -231,13 +231,13 @@ def __eq__(self, other):
...
 
-    # TODO: RUSTPYTHON
-    @unittest.expectedFailure
+    # # TODO: RUSTPYTHON
+    # @unittest.expectedFailure
     def test_contains_fake(self):  # XXX: RUSTPYTHON; the method also need to be removed when done
         super().test_contains_fake()
```
Comment (fanninpm) [inline Lib/test/test_list.py]:
```suggestion
```

**Review by fanninpm — CHANGES_REQUESTED** [review]
Please don't leave these things behind in the code. They're not there in CPython.

**Review by fanninpm — COMMENTED** [review]
LGTM.

**Comment by youknowone** [comment]
@qingshi163 could you review this PR?

**Comment by qingshi163** [comment]
As I remember, the guard or borrower is to optimize the compare by check if the class is not the heap type, so we can reuse the compare function in each loop. Also to use rich_compare_bool we cannot reuse the PyTypeRef for the compared object but this will be no more the issue if we can fetch the PyType atomicly in the future.

Yes, the correctness is more important than performance, but if you can solve the problem without drawback the performance could be amazing. if not please advise the issue or the failing point of the current implement, so we could see why this is not going to work.

@youknowone 

Thanks

**Review by youknowone — APPROVED** [review]
@qingshi163 I created an issue to refer this pr #4008
@TwoPair Thank you!

**Pull request merged**